megatron/mpu/layers.py

```
@@ -217,6 +217,9 @@ def __init__(self, num_embeddings, embedding_dim,
 
 
     def forward(self, input_):
+        if torch.any(input_ >= self.num_embeddings):
+            raise ValueError("There is an input id in the input that is greater than the highest possible input id.")
+
```

killing the training at run time because the input is broken? That we can't support.

Additionally this error can't be acted upon since the operator would have no idea how to fix this as you're not including the input id and the same id - how can this get fixed?

If there is a need to validate data before the training it should happen separately from the training.

---

the forward should be as lean as possible and not need to check anything so that it could run as fast as possible.

---

I don't agree, we need this sanity check so that we're not doing something wrong. Right now, model would silently bypass this issue if you use TP>2. IMO it should kill the training as we are doing something VERY bad.

---

You don't do data sanity checks in `forward`. All sanity checks should ideally be done before at the dataloader level.

Which case are you guarding against - all of the data is completely borked or only some inputs are invalid?

---

and practically, let's do a an imaginary scenario - you started the training and this assert happens 5 days in - what do you do? I fail to see how this assert is actionable.

---

It will crash with the cross entropy loss no?

---

We're trying to catch if some inputs are invalid. 

> and practically, let's do a an imaginary scenario - you started the training and this assert happens 5 days in - what do you do? I fail to see how this assert is actionable.

The same we would do if we ran in this exception using TP=1. We manually re-process the data, or try to remove that sample from the dataset ... We have this assert happen in TP=1 right now, so it makes sense not to have silent errors in TP=2.

---

how will you find this sample?

why force crash the program and force a huge down time until you figure it out if there was a mistake in a single input? 

why not move it into dataloader and just skip it and possibly warn about it in case it's not a single bad input but a lot of it?

---

@Stas, just for information we also modify a little bit the error message in order to log also the input ids and the embedding size

---

Initially, this test was motivated by the fact that we wanted to be sure that the embedding matrix created with the tensor parallelism feature is the desired one. And we realized that contrary to TP=1, this type of error would pass without raising an error.

In my opinion, we can also add a test on the data side, but I think that this error is still relevant as we need to be sure that the embedding matrix is in agreement with the tokenizer we use. 

If you ever see another way to do it at the time of the creation of the model sharded architecture on the gpu's, it would be better but personally I don't see any place that would allow to do that since the "routing" of the input ids is done in the forward.

---

Usually what happens if you index out of bounds is a completely undecipherable cuda error at a later point. It is definitely better to include this assert. All it does is improve the error message.

Another alternative I guess, if we really want to continue despite there being a bug is to send a warning and do `input_=min(input, self.num_embeddings)`, though I think I prefer the assert as we can't miss it and we should probably fix and restart it anyway if this happens.

---

sorry, unfortunately I don't have the resources to dive right now into this as I have to finish lots of things before the launch, so I trust you will do the best thing you can and if things break during the training I will ping you and you will know what to do.

It's not great to have this sort of last minute change that hasn't been thoroughly tested but what to do.

---

Thomas and I discussed this, he helped me to clear out an important misunderstanding and he will post an update.

Thank you for working on this, @SaulLu and @DanielHesslow - my apologies that I can't be involved at a deeper level at the moment.

---

Okay talked to @stas00 :
 - This PR makes sense in the sense that the model should check for this. @DanielHesslow said it better, but if we get a out of bound index, we need to raise an exception
 - There's the issue of "if the training set is bad for X reason" what do we do? After this PR it will throw badly. So we need a strategy. Given that the number of root causes, @stas00 suggest to have a skip mechanism in the data loader, why? Because we're able to keep training, in case the issue "isn't that bad". In the worst case scenario, we kill the job and relaunch from scratch. But this is orthogonal to this PR IMO.

---

yes, so the idea is do a check on the dataloader level, and warn when skipping a bad sample. If it's a lot of samples but not the majority - we continue training while someone is fixing the data.

of course if the data is very broken then it'll be skipping them all and we can't train.

But the point is that - don't crash the training unless you have to.

---

and additionally **we can** pre-check all our data outside of the training so that it doesn't have any bad samples.

This PR is mainly for future users, but in our training we should make sure it never hits this assert